sumo/src/normalize.rs

```diff
@@ -0,0 +1,149 @@
+//! Transforms a `raw::Network` into a `Network` that's easier to reason about.
+
+use std::collections::BTreeMap;
+use std::error::Error;
+
+use abstutil::Timer;
+use geom::{Distance, PolyLine, Pt2D, Ring};
+
+use crate::{raw, Edge, InternalEdge, InternalLane, Junction, Lane, Network};
+
+impl Network {
+    /// Reads a .net.xml file and return the normalized SUMO network.
+    pub fn load(path: &str, timer: &mut Timer) -> Result<Network, Box<dyn Error>> {
+        let raw = raw::Network::parse(path, timer)?;
+        timer.start("normalize");
+        let network = Network::from_raw(raw);
+        timer.stop("normalize");
+        Ok(network)
+    }
+
+    fn from_raw(raw: raw::Network) -> Network {
+        let mut network = Network {
+            location: raw.location,
+            normal_edges: BTreeMap::new(),
+            internal_edges: BTreeMap::new(),
+            junctions: BTreeMap::new(),
+            connections: raw.connections,
+        };
+
+        let types: BTreeMap<String, raw::Type> =
+            raw.types.into_iter().map(|t| (t.id.clone(), t)).collect();
+
+        for junction in raw.junctions {
+            if junction.junction_type == "internal" {
+                continue;
+            }
+            network.junctions.insert(
+                junction.id.clone(),
+                Junction {
+                    pt: junction.pt(),
+                    id: junction.id,
+                    junction_type: junction.junction_type,
+                    incoming_lanes: junction.incoming_lanes,
+                    internal_lanes: junction.internal_lanes,
+                    shape: junction.shape.unwrap(),
+                },
+            );
+        }
+
+        for edge in raw.edges {
+            if edge.function == raw::Function::Internal {
+                let mut lanes = Vec::new();
+                for lane in edge.lanes {
+                    lanes.push(InternalLane {
+                        id: lane.id,
+                        index: lane.index,
+                        speed: lane.speed,
+                        length: lane.length,
+                        center_line: lane.shape.ok(),
+                        allow: lane.allow,
+                    });
+                }
+                network
+                    .internal_edges
+                    .insert(edge.id.clone(), InternalEdge { id: edge.id, lanes });
+                continue;
+            }
```

Comment:
else we don't care about the value of `function`? Or are the only two values "Internal" and "normal"?

Reply:
In pedesrian networks you will also see the values "crossing" and "walkingare". When in doubt, refer to  https://github.com/eclipse/sumo/blob/master/data/xsd/net_file.xsd#L53